File: Threaded Binary Trees/threaded_binary_tree_operations.py
# File: threaded_binary_tree.py

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None
        self.lThread = True
        self.rThread = True


class ThreadedBinaryTree:

    def __init__(self):
        self.root = None

    def insert(self, key):
        parent = None
        current = self.root

        while current:
            if key == current.data:
                return

            parent = current

            if key < current.data:
                if not current.lThread:
                    current = current.left
                else:
                    break
            else:
                if not current.rThread:
                    current = current.right
                else:
                    break

        node = Node(key)

        if parent is None:
            self.root = node
        elif key < parent.data:
            node.left = parent.left
            node.right = parent
            parent.lThread = False
            parent.left = node
        else:
            node.left = parent
            node.right = parent.right
            parent.rThread = False
            parent.right = node

    def inorder(self):
        current = self.root
        while current and not current.lThread:
            current = current.left

        while current:
            print(current.data, end=" ")
            if current.rThread:
                current = current.right
            else:
                current = current.right
                while current and not current.lThread:
                    current = current.left

    def search(self, key):
        current = self.root
        while current:
            if key == current.data:
                return current
            if key < current.data:
                if not current.lThread:
                    current = current.left
                else:
                    break
            else:
                if not current.rThread:
                    current = current.right
                else:
                    break
        return None


if __name__ == "__main__":
    tree = ThreadedBinaryTree()

    tree.insert(20)
    tree.insert(10)
    tree.insert(30)
    tree.insert(25)

    print("Inorder Traversal:")
    tree.inorder()

File: Threaded Binary Trees/ThreadedBinaryTreeOperations.java
// File: ThreadedBinaryTree.java

class Node {
    int data;
    Node left, right;
    boolean lThread, rThread;

    Node(int data) {
        this.data = data;
        lThread = rThread = true;
    }
}

public class ThreadedBinaryTree {

    Node root;

    public void insert(int key) {
        Node parent = null, current = root;

        while (current != null) {
            if (key == current.data)
                return;

            parent = current;

            if (key < current.data) {
                if (!current.lThread)
                    current = current.left;
                else
                    break;
            } else {
                if (!current.rThread)
                    current = current.right;
                else
                    break;
            }
        }

        Node node = new Node(key);

        if (parent == null)
            root = node;
        else if (key < parent.data) {
            node.left = parent.left;
            node.right = parent;
            parent.lThread = false;
            parent.left = node;
        } else {
            node.left = parent;
            node.right = parent.right;
            parent.rThread = false;
            parent.right = node;
        }
    }

    public void inorder() {
        Node current = root;
        while (current != null && !current.lThread)
            current = current.left;

        while (current != null) {
            System.out.print(current.data + " ");
            if (current.rThread)
                current = current.right;
            else {
                current = current.right;
                while (current != null && !current.lThread)
                    current = current.left;
            }
        }
    }

    public Node search(int key) {
        Node current = root;

        while (current != null) {
            if (key == current.data)
                return current;

            if (key < current.data) {
                if (!current.lThread)
                    current = current.left;
                else
                    break;
            } else {
                if (!current.rThread)
                    current = current.right;
                else
                    break;
            }
        }
        return null;
    }

    public static void main(String[] args) {
        ThreadedBinaryTree tree = new ThreadedBinaryTree();

        tree.insert(20);
        tree.insert(10);
        tree.insert(30);
        tree.insert(25);

        System.out.println("Inorder Traversal:");
        tree.inorder();
    }
}
